File: lib/features/nutrition/data/wear_sync_contract.dart
import 'dart:async' show unawaited;
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';

const int wearSyncSchemaVersion = 2;
const String wearSyncOriginPhone = 'phone';
const String wearSyncOriginWatch = 'watch';
const String wearSyncEntityActiveWorkout = 'active_workout';
const String wearSyncEntityFasting = 'fasting';

class WearSyncEnvelope {
  const WearSyncEnvelope({
    required this.schemaVersion,
    required this.entity,
    required this.entityId,
    required this.revision,
    required this.origin,
    required this.updatedAtEpochMs,
    required this.payload,
  });

  final int schemaVersion;
  final String entity;
  final String entityId;
  final int revision;
  final String origin;
  final int updatedAtEpochMs;
  final Map<String, dynamic> payload;

  Map<String, dynamic> toJson() {
    return {
      'schemaVersion': schemaVersion,
      'entity': entity,
      'entityId': entityId,
      'revision': revision,
      'origin': origin,
      'updatedAtEpochMs': updatedAtEpochMs,
      'payload': payload,
    };
  }

  String encode() => jsonEncode(toJson());

  static WearSyncEnvelope wrap({
    required String entity,
    required String entityId,
    required int revision,
    required String origin,
    required Map<String, dynamic> payload,
    int? updatedAtEpochMs,
  }) {
    return WearSyncEnvelope(
      schemaVersion: wearSyncSchemaVersion,
      entity: entity,
      entityId: entityId,
      revision: revision,
      origin: origin,
      updatedAtEpochMs:
          updatedAtEpochMs ?? DateTime.now().millisecondsSinceEpoch,
      payload: payload,
    );
  }

  static WearSyncEnvelope decode(
    String jsonString, {
    required String fallbackEntity,
    required String fallbackEntityId,
    required String fallbackOrigin,
  }) {
    final decoded = jsonDecode(jsonString) as Map<String, dynamic>;
    final payload = decoded['payload'];
    // origin/entityId are optional at decode time (the fallback below still
    // covers "absent") but must be the right type when present — a
    // wrong-typed value falls through to the legacy branch instead of
    // crashing or being silently coerced.
    final isEnvelope =
        decoded['schemaVersion'] is num &&
        decoded['entity'] is String &&
        payload is Map &&
        (decoded['origin'] == null || decoded['origin'] is String) &&
        (decoded['entityId'] == null || decoded['entityId'] is String);

    if (!isEnvelope) {
      return WearSyncEnvelope(
        schemaVersion: 0,
        entity: fallbackEntity,
        entityId: fallbackEntityId,
        revision: (decoded['revision'] as num?)?.toInt() ?? 0,
        origin: fallbackOrigin,
        updatedAtEpochMs:
            (decoded['updatedAtEpochMs'] as num?)?.toInt() ??
            (decoded['startedAtEpochMs'] as num?)?.toInt() ??
            0,
        payload: decoded,
      );
    }

    return WearSyncEnvelope(
      schemaVersion: (decoded['schemaVersion'] as num).toInt(),
      entity: decoded['entity'] as String,
      entityId: decoded['entityId'] as String? ?? fallbackEntityId,
      revision: (decoded['revision'] as num?)?.toInt() ?? 0,
      origin: decoded['origin'] as String? ?? fallbackOrigin,
      updatedAtEpochMs: (decoded['updatedAtEpochMs'] as num?)?.toInt() ?? 0,
      payload: Map<String, dynamic>.from(payload),
    );
  }
}

/// Issues revisions that never go backwards, even across a process restart.
///
/// The old in-memory-only version reset to 0 on every app restart, so a
/// revision issued before a restart could collide with (or beat) one issued
/// after it — see Phase 1 of
/// docs/wear-sync-race-conditions-remediation-plan-2026-08-11.md. [key]
/// namespaces the persisted high-water mark so independent entities (e.g.
/// workout vs. fasting) don't share — and corrupt — each other's sequence.
class WearRevisionAllocator {
  WearRevisionAllocator(this._prefs, String key) : _key = 'wear_sync_revision_$key';

  final SharedPreferences _prefs;
  final String _key;

  int next() {
    final persistedLast = _prefs.getInt(_key) ?? 0;
    final now = DateTime.now().millisecondsSinceEpoch;
    final next = now > persistedLast ? now : persistedLast + 1;
    // shared_preferences updates its synchronous in-memory cache immediately
    // on setInt — the returned Future only tracks the async platform-channel
    // flush to disk, so a subsequent next() call in this process sees `next`
    // right away even though this call isn't awaited. Awaiting here would
    // make every sync call chain through disk I/O for no correctness gain.
    unawaited(_prefs.setInt(_key, next));
    return next;
  }
}

class WearDedupeState {
  final Map<String, ({int revision, int updatedAtEpochMs})> _latest = {};

  /// Pure check: would [envelope] be newer than the last envelope *committed*
  /// for its key? Does not mutate any state.
  ///
  /// Split from the combined check-and-commit that used to live here so a
  /// caller can gate on this *before* attempting to durably apply the
  /// envelope, then call [commit] only once that apply actually succeeds. The
  /// combined version committed on the check alone, so a failed apply (e.g. a
  /// Drift write that threw) still marked the revision "seen" and a retried
  /// delivery of the same envelope was then silently dropped forever — see
  /// docs/wear-sync-race-conditions-remediation-plan-2026-08-11.md Phase 2
  /// (ENG-07/10/12).
  ///
  /// The unenveloped-legacy shape is rejected (fail-closed) rather than
  /// always-accepted — see Phase 1c of the remediation plan. That bypass
  /// existed for pre-envelope watch builds; once both APKs move to
  /// schemaVersion 2 together there's no legitimate sender left for it.
  bool wouldAccept(WearSyncEnvelope envelope) {
    if (_isUnenvelopedLegacyPayload(envelope)) return false;
    final previous = _latest[_keyFor(envelope)];
    if (previous == null) return true;
    if (envelope.revision < previous.revision) return false;
    if (envelope.revision == previous.revision &&
        envelope.updatedAtEpochMs <= previous.updatedAtEpochMs) {
      return false;
    }
    return true;
  }

  /// Records [envelope] as applied. Call only after the corresponding
  /// durable write has succeeded.
  void commit(WearSyncEnvelope envelope) {
    if (_isUnenvelopedLegacyPayload(envelope)) return;
    _latest[_keyFor(envelope)] = (
      revision: envelope.revision,
      updatedAtEpochMs: envelope.updatedAtEpochMs,
    );
  }

  bool _isUnenvelopedLegacyPayload(WearSyncEnvelope envelope) =>
      envelope.schemaVersion == 0 &&
      envelope.revision == 0 &&
      envelope.updatedAtEpochMs == 0;

  String _keyFor(WearSyncEnvelope envelope) =>
      '${envelope.entity}:${envelope.entityId}:${envelope.origin}';
}

String normalizeWearSetType(String? raw) {
  switch ((raw ?? 'standard').trim().toLowerCase()) {
    case 'normal':
    case 'work':
    case 'working':
    case '':
      return 'standard';
    case 'warmup':
    case 'warm_up':
      return 'standard';
    case 'failure':
    case 'forced':
      return 'forced';
    default:
      return raw!.trim().toLowerCase();
  }
}

bool normalizeWearWarmup({String? setType, bool? isWarmup}) {
  final raw = (setType ?? '').trim().toLowerCase();
  return isWarmup == true || raw == 'warmup' || raw == 'warm_up';
}

File: android/wear/src/main/java/com/ams/herculex/sync/WearSyncContract.kt
package com.ams.herculex.sync

import android.content.Context
import android.util.Log
import org.json.JSONObject

data class SyncEnvelope(
    val schemaVersion: Int,
    val entity: String,
    val entityId: String,
    val revision: Long,
    val origin: String,
    val updatedAtEpochMs: Long,
    val payload: JSONObject,
)

object WearSyncContract {
    const val SCHEMA_VERSION = 2
    const val ENTITY_ACTIVE_WORKOUT = "active_workout"
    const val ENTITY_FASTING = "fasting"
    const val ORIGIN_PHONE = "phone"
    const val ORIGIN_WATCH = "watch"

    fun decodeEnvelope(
        json: String,
        fallbackEntity: String,
        fallbackEntityId: String,
        fallbackOrigin: String,
    ): SyncEnvelope {
        val obj = JSONObject(json)
        val payload = obj.optJSONObject("payload")
        return if (payload != null && obj.has("schemaVersion") && obj.has("entity")) {
            SyncEnvelope(
                schemaVersion = obj.optInt("schemaVersion", SCHEMA_VERSION),
                entity = obj.optString("entity", fallbackEntity),
                entityId = obj.optString("entityId", fallbackEntityId),
                revision = obj.optLong("revision", 0L),
                origin = obj.optString("origin", fallbackOrigin),
                updatedAtEpochMs = obj.optLong("updatedAtEpochMs", 0L),
                payload = payload,
            )
        } else {
            SyncEnvelope(
                schemaVersion = 0,
                entity = fallbackEntity,
                entityId = fallbackEntityId,
                revision = obj.optLong("revision", 0L),
                origin = fallbackOrigin,
                updatedAtEpochMs = obj.optLong("updatedAtEpochMs", obj.optLong("startedAtEpochMs", 0L)),
                payload = obj,
            )
        }
    }

    fun encodeEnvelope(
        entity: String,
        entityId: String,
        revision: Long,
        origin: String,
        payload: JSONObject,
    ): String {
        return JSONObject()
            .put("schemaVersion", SCHEMA_VERSION)
            .put("entity", entity)
            .put("entityId", entityId)
            .put("revision", revision)
            .put("origin", origin)
            .put("updatedAtEpochMs", System.currentTimeMillis())
            .put("payload", payload)
            .toString()
    }

    fun normalizeSetType(raw: String?): String {
        return when (raw?.trim()?.lowercase().orEmpty()) {
            "", "normal", "work", "working" -> "standard"
            "warmup", "warm_up" -> "standard"
            "failure" -> "forced"
            else -> raw!!.trim().lowercase()
        }
    }

    fun normalizeIsWarmup(setType: String?, isWarmup: Boolean): Boolean {
        val raw = setType?.trim()?.lowercase().orEmpty()
        return isWarmup || raw == "warmup" || raw == "warm_up"
    }
}

/// Issues revisions that never go backwards, even across a process restart.
///
/// Unlike the Dart side (`WearRevisionAllocator` in wear_sync_contract.dart),
/// there was no revision-allocator construct here at all before Phase 1 —
/// callers just inlined `System.currentTimeMillis()`, which reset on every
/// restart. [key] namespaces the persisted high-water mark per entity (e.g.
/// "workout" vs. "fasting") so their sequences can't collide. See Phase 1 of
/// docs/wear-sync-race-conditions-remediation-plan-2026-08-11.md.
class WearRevisionAllocator(context: Context, key: String) {
    private val prefs = context.applicationContext
        .getSharedPreferences("wear_sync_revision", Context.MODE_PRIVATE)
    private val prefKey = "revision_$key"

    @Synchronized
    fun next(): Long {
        val persistedLast = prefs.getLong(prefKey, 0L)
        val now = System.currentTimeMillis()
        val next = if (now > persistedLast) now else persistedLast + 1
        prefs.edit().putLong(prefKey, next).apply()
        return next
    }
}

/// Split (Phase 4) from a single combined `shouldAccept` into a pure
/// [wouldAccept] check and a mutating [commit], mirroring the Dart-side
/// `WearDedupeState.wouldAccept`/`commit` split from Phase 2. The old
/// combined method wrote the high-water mark as a side effect of the check
/// itself, *before* the caller's durable write (`persistSession`/
/// `WorkoutStore.parseAndUpdateSession`/`FastingStore.saveSnapshot`) had even
/// been attempted — a failed write still poisoned the dedupe state, so a
/// retry of the identical envelope was silently dropped forever. This was
/// deferred from both Phase 1 and Phase 2 (see their progress-log entries).
class AppliedRevisionStore(context: Context) {
    private val prefs = context.applicationContext
        .getSharedPreferences("wear_sync_applied_revisions", Context.MODE_PRIVATE)

    /// Pure check — does not mutate the persisted high-water mark.
    fun wouldAccept(envelope: SyncEnvelope): Boolean {
        // Fail-closed (Phase 1c): this bypass existed for pre-envelope watch
        // builds. Once both APKs move to schemaVersion 2 together there's no
        // legitimate sender left for this shape, so reject instead of
        // always-accepting it.
        if (envelope.schemaVersion == 0 && envelope.revision == 0L && envelope.updatedAtEpochMs == 0L) {
            return false
        }
        val key = "${envelope.entity}:${envelope.entityId}:${envelope.origin}"
        val previousRevision = prefs.getLong("${key}:revision", Long.MIN_VALUE)
        val previousUpdatedAt = prefs.getLong("${key}:updatedAt", Long.MIN_VALUE)
        if (envelope.revision < previousRevision) return false
        if (envelope.revision == previousRevision && envelope.updatedAtEpochMs <= previousUpdatedAt) {
            return false
        }
        return true
    }

    /// Mutation only — call after the corresponding durable write succeeds.
    fun commit(envelope: SyncEnvelope) {
        val key = "${envelope.entity}:${envelope.entityId}:${envelope.origin}"
        prefs.edit()
            .putLong("${key}:revision", envelope.revision)
            .putLong("${key}:updatedAt", envelope.updatedAtEpochMs)
            .apply()
    }
}

fun logWearSync(envelope: SyncEnvelope, path: String, delivery: String, apply: String) {
    Log.d(
        "WearSync",
        "entity=${envelope.entity} revision=${envelope.revision} origin=${envelope.origin} " +
            "entityId=${envelope.entityId} path=$path delivery=$delivery apply=$apply",
    )
}
